files/src/hobby_hub_utils/hobby_hub_pkg/pin_manager.py

```diff
@@ -0,0 +1,299 @@
+    """Return the value associated with the key in the json file.
+
+    Args:
+        tag (str): tag to search for
+
+    Returns:
+        str: value associated with tag
+    """
+    logging.info(f'Getting pin {tag}')
+    value = None
+    with open(pin_config_filename, 'r+') as f:
+        pin_config = json.load(f)
+        if tag in pin_config:
+            value = pin_config[tag]
+            logging.info(f'{tag}:{value} found in JSON file')
+            pin_config[tag]["in_use"] = True
+            f.seek(0) # should reset file position to the beginning.
+            json.dump(pin_config, f, indent=4)
+            f.truncate() # remove remaining part
+            logging.info(f'{tag} marked as in use')
+
+    return value
+
+
+def reset(pin_config_filename):
+    """Resets the pin configuration JSON file"""
+    logging.info('Resetting config')
+    with open(pin_config_filename, 'w') as f:
+        f.write("{}")
+        logging.info('Config reset')
+
+
+def clear_unused(pin_config_filename):
+    """Identify and remove unused key-value pairs
+
+    Only the pins which are requested from the running programs should be available in the JSON file
+    """
+    logging.info('Clearing unused pins')
+    with open(pin_config_filename, 'r+') as f:
+        pin_config = json.load(f)
+        unused_pins = [tag for tag in pin_config if pin_config[tag]["in_use"] == False]
+        for tag in unused_pins:
+            logging.info(f'Removing unused pin {tag}')
+            del pin_config[tag]
+        f.seek(0) # should reset file position to the beginning.
+        json.dump(pin_config, f, indent=4)
+        f.truncate() # remove remaining part
+        logging.info('All unused pins cleared')
+
+
+def request_pin(tag, typ):
+    with open(PIN_MAP_FILE, 'r+') as f:
+        pin_config = json.load(f)
+        physical_pin = ""
+        if tag in pin_config:
+            if pin_config[tag]["type"] == typ:
+                physical_pin = pin_config[tag]["pin"]
+            else:
+                raise KeyError("Requested pin " + tag + " of type: " + type + " already found as type: " + pin_config[tag]["type"])
+        else: # TODO fix this logic
+            used_pins = []
+            for config in pin_config.values():
+                used_pins.append(config["pin"])
+            for pin in pin_table:
+                if (pin[0] not in used_pins) and (pin[2] == typ):
+                    pin_config[tag] = {}
+                    pin_config[tag]["pin"] = pin[0]
+                    pin_config[tag]["type" ] = typ
+                    physical_pin = pin[0]
+                    break
+            f.seek(0) # should reset file position to the beginning.
+            json.dump(pin_config, f, indent=4)
+            f.truncate() # remove remaining part
+            logging.info(f'{tag} added to JSON with placeholder')
+
+        return physical_pin
+
+
+def get_gpio(tag):
+    return Pin(request_pin(tag, Pin_t.GPIO))
+
+
+def get_led(tag):
```

Comment:
could you add some documentation to the methods the user should be using? in the code and probobly a usage guide on the wiki?

Reply:
> Also I checked out your branch, ran the setup script and rebooted the board. when i logged back in, the /etc/hobby-hub/pin_mapping.json was unedited. isn't that supposed to change? also i'm getting 2 blinking led's but they are blinking very erratically and much quicker with less opacity. Not the same as the previous test programs.

Change to what? Should the json file be cleared on startup?

Reply:
> could you add some documentation to the methods the user should be using? in the code and probobly a usage guide on the wiki?

Sure

Reply:
doesn't the JSON reflect the current pins in use? since that's what barty is reading.